### PR TITLE
Fix for Git plugin calls fetch twice per checkout

### PR DESCRIPTION
**Pull request by Loghijiaha**
## [JENKINS-49757](https://issues.jenkins-ci.org/browse/JENKINS-49757) - Redundant Git fetch has resolved

After removing fetch.execute, testRedundantFetch passes successfully.

`Building in workspace /home/loghi/Documents/GSoC/git-plugin/work/workspace/fetch_twice_fixed
No credentials specified
Cloning the remote Git repository
Cloning repository https://github.com/Loghijiaha/ipython
 > git init /home/loghi/Documents/GSoC/git-plugin/work/workspace/fetch_twice_fixed # timeout=10
Fetching upstream changes from https://github.com/Loghijiaha/ipython
 > git --version # timeout=10
 > git fetch --tags --force --progress -- https://github.com/Loghijiaha/ipython +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/Loghijiaha/ipython # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/Loghijiaha/ipython # timeout=10
 > git rev-parse refs/remotes/origin/master^{commit} # timeout=10
 > git rev-parse refs/remotes/origin/origin/master^{commit} # timeout=10
Checking out Revision 1f372309c074c62d1e13aeebf33c60726a9baff4 (refs/remotes/origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 1f372309c074c62d1e13aeebf33c60726a9baff4 # timeout=10
Commit message: "Finally done with simple plugin for ipython"
First time build. Skipping changelog.
Finished: SUCCESS`

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] Dependency or infrastructure update
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 1]
```diff
@@ -904,7 +904,6 @@ private void fetchFrom(GitClient git,
                 for (GitSCMExtension extension : extensions) {
                     extension.decorateFetchCommand(this, run, git, listener, fetch);
                 }
-                fetch.execute();
```
Comment (rishabhBudhouliya) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Upon the removal of fetch.execute(), if `git.hasGitRepo()` is true which means that we are "Fetching changes from the remote Git repository", how would GitSCM _retrieve_ the changes in the remote repo?

File: src/test/java/hudson/plugins/git/GitSCMTest.java [hunk 2]
```diff
@@ -143,6 +149,30 @@ private StandardCredentials getInvalidCredential() {
         return new UsernamePasswordCredentialsImpl(scope, id, "desc: " + id, username, password);
     }
 
+    @Issue("JENKINS-56404")
+    @Test
+    public void testRedundantFetch() throws Exception {
```
Comment (rishabhBudhouliya) [inline src/test/java/hudson/plugins/git/GitSCMTest.java]: While the first time checkout would work, could you write a case which tests whether `retrieveChanges` is able to fetch an update?

**Comment by MarkEWaite** [comment]
I won't be able to review this pull request or #845 until next week and there is a risk it will be delayed one more week since I am traveling next week as part of the Southern California Linux Expo conference.

I suspect neither of the pull requests have evaluated the subtle problems that are related to refspecs in the potentially redundant fetch commands.  The cases that I recognize include:

| First Fetch     | Second Fetch      | Safe to Skip Second Fetch?    |
| -------------- | ------------------ | ------------------------------    |
| Any               | Same as first       |  Yes                                          |
| Narrow         | Wide                   | No - wide refspec brings data |
| Wide             | Narrow               | Yes if narrow is subset of wide |
| Disjoint-1      | Disjoint-2           | No - disjoint-2 brings data      |

When I say "narrow", I mean a refspec like:
```
+refs/heads/stable-3.x:refs/remotes/origin/stable-3.x
```

When I say "wide", I mean a refspec like:

```
+refs/heads/*:refs/remotes/origin/*
```

When I say "disjoiont", I mean refspecs that refer to separate trees like:

```
+refs/heads/stable-3.x:refs/remotes/origin/stable-3.x
+refs/heads/stable-2.x:refs/remotes/origin/stable-2.x
```

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite , fair enough. It seems that there is a need for a deeper inspection and what you've suggested is going to help a lot. Thanks!

**Comment by Loghijiaha** [comment]
> I won't be able to review this pull request or #845 until next week and there is a risk it will be delayed one more week since I am traveling next week as part of the Southern California Linux Expo conference.
> 
> I suspect neither of the pull requests have evaluated the subtle problems that are related to refspecs in the potentially redundant fetch commands. The cases that I recognize include:
> First Fetch 	Second Fetch 	Safe to Skip Second Fetch?
> Any 	Same as first 	Yes
> Narrow 	Wide 	No - wide refspec brings data
> Wide 	Narrow 	Yes if narrow is subset of wide
> Wide 	Narrow 	No if narrow is not a subset of wide
> Disjoint-1 	Disjoint-2 	No - disjoint-2 brings data
> 
> When I say "narrow", I mean a refspec like:
> 
> ```
> +refs/heads/stable-3.x:refs/remotes/origin/stable-3.x
> ```
> 
> When I say "wide", I mean a refspec like:
> 
> ```
> +refs/heads/*:refs/remotes/origin/*
> ```
> 
> When I say "disjoiont", I mean refspecs that refer to separate trees like:
> 
> ```
> +refs/heads/stable-3.x:refs/remotes/origin/stable-3.x
> +refs/heads/stable-2.x:refs/remotes/origin/stable-2.x
> ```
> 
> I will want tests which confirm those cases are handled. If we don't have tests that confirm those cases are handled, we risk breaking users with an optimization that inadvertently reduces the amount of repository information that is fetched into the workspace.

**Comment by Loghijiaha** [comment]
@MarkEWaite Thank you for the  suggestions.

File: src/main/java/hudson/plugins/git/GitSCM.java (hunk 1 above)
Comment (Loghijiaha) [reply src/main/java/hudson/plugins/git/GitSCM.java]: I would have checked other cases also. I will resubmit some changes.

File: src/test/java/hudson/plugins/git/GitSCMTest.java (hunk 2 above)
Comment (Loghijiaha) [reply src/test/java/hudson/plugins/git/GitSCMTest.java]: I will try to do it after the changes made.

**Comment by rishabhBudhouliya** [comment]
@MarkEWaite I have failed to understand the scenario where we would have the case 

Narrow | Wide | No - wide refspec brings data
-- | -- | --

If we do not honor refspec for initial clone, we fetch all branches. In what case would we have a narrow refspec in the initial fetch call?

**Comment by MarkEWaite** [comment]
> If we do not honor refspec for initial clone, we fetch all branches. In what case would we have a narrow refspec in the initial fetch call?

I think you're correct.  I hadn't considered that the default refspec is wide unless we honor refspec on initial clone.  When we honor refspec on initial clone, the refspec is the same in both fetches.  Thanks for catching my mistake!

**Comment by MarkEWaite** [comment]
Closing this pull request in favor of #845 and #904 .  Thanks for starting the efforts @Loghijiaha !

**Pull request closed without merging**